### DISCUSSION TITLE
Warm starts from MiniZinc

### DISCUSSION THREAD

**Discussion by glebbelov** [google/or-tools#2999] [CP-SAT questions] [answered]
Hi,

warm start annotations are being added in MiniZinc. I wonder if your flatzinc CP solver can support them and if you wish to extend the annotations. To be used, e.g., in the sunny-cp framework or in iterative approaches. For details, see README_MIP.txt on the develop branch, currently supported only for Gurobi and CPLEX backends.

Example:

```
array[1..3] of var 0..10: x;
array[1..3] of var 0.0..10.5: xf;
var bool: b;
array[1..3] of var set of 5..9: xs;
constraint b+sum(x)==1;
constraint b+sum(xf)==2.4;
constraint 5==sum( [ card(xs[i]) | i in index_set(xs) ] );
solve
  :: warm_start( [b], [<>] )                         %%% Use <> for missing values
  :: warm_start_array( [
       warm_start( x, [<>,8,4] ),
       warm_start( xf, array1d(-5..-3, [5.6,<>,4.7] ) ),
       warm_start( xs, array1d( -3..-2, [ 6..8, 5..7 ] ) )
     ] )
  :: seq_search( [ int_search(x, first_fail, indomain_min, complete)  ] )
  minimize x[1] + b + xf[2] + card( xs[1] intersect xs[3] );

```

**Comment by lperron** [comment]
The Flatzinc CP is under deep maintenance :-)

We may look at it for the SAT backend.

Thanks

Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00


2017-12-06 5:33 GMT+01:00 glebbelov <notifications@github.com>:

**Comment by lperron** [comment]
I tried to compile the example.
It crashes with minizinc 2.2.1 :-(

**Comment by glebbelov** [comment]
Cannot reproduce the issue. (This is not to bother you on weekend!)

The example flattens well on KUbuntu 16.04, for std, linear, and
or-tools-sat libraries (current, 2b1d70ac). However fzn-or-tools has two
things: the float array xf needs to be removed, but then it fails to parse
fzn because
  :: warm_start( [b], [<>] ) %%% Use <> for missing values
has been flattened to empty annotation:
  :: warm_start([],[])



On Fri, 7 Sep 2018 at 18:50, Laurent Perron <notifications@github.com>
wrote:

> I tried to compile the example.
> It crashes with minizinc 2.2.1 :-(
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/or-tools/issues/539#issuecomment-419369959>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AQgXZpIEQMFqv8Umb-UvxVj4ms-jFYEMks5uYjNQgaJpZM4Q3W-k>
> .
>


-- 
Dr Gleb Belov       Monash University      +61 3 9903 1622

**Comment by lalithsuresh** [comment]
Any plans to support warm_start annotations for the CP-SAT backend? I'd be happy to test this out if there's any early code available for this.

**Comment by lperron** [comment]
I just need the spec of the flatzinc format, and some tests :-)  plus some
time.
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00



Le jeu. 21 févr. 2019 à 04:55, Lalith Suresh <notifications@github.com> a
écrit :

**Comment by glebbelov** [comment]
Hi,

Flatzinc format is whatever you redefine (or not, then you get what is
declared in stdlib.mzn). For example, std/nosets.mzn already redefines
warm_start for set variables into something using booleans:

annotation warm_start( array[int] of var set of int: x, array[int] of set
of int: v ) =
   warm_start_array( [
       let {
         array[int] of var bool: xb = set2bools(x[i]),
         set of int: is_var = ub(x[i]) diff lb(x[i]),
         int: iV = i - min(index_set(x)) + if 0<length(v) then
min(index_set(v)) else 0 endif,
       } in
       warm_start( [ xb[k] | k in is_var ],
         if iV in index_set(v) then [ k in v[iV] | k in is_var ] else []
endif )
   | i in index_set(x) ] );       %% ignoring v[i] diff ub(x[i]) and
lb(x[i]) diff v[i]

All you need to do then, similarly as you process search annotations,
process the warm_starts. The original declarations for the warm_start
annotations are in stdlib.mzn, as well as redefinitions for optional
variables, so anything you don't redefine can be passed to FZN. stdlib.mzn
also contains the remark that the order of warm_starts, relative to other
search annotations, can be important, so they all might need to be put into
a seq_search as below.

The example I used for testing (currently in README_MIP.txt but to be moved
to documentation):
array[1..3] of var 0..10: x;
array[1..3] of var 0.0..10.5: xf;
var bool: b;
array[1..3] of var set of 5..9: xs;
constraint b+sum(x)==1;
constraint b+sum(xf)==2.4;
constraint 5==sum( [ card(xs[i]) | i in index_set(xs) ] );
solve
  :: warm_start_array( [                     %%% Can be on the upper level
    warm_start( x, [<>,8,4] ),               %%% Use <> for missing values
    warm_start( xf, array1d(-5..-3, [5.6,<>,4.7] ) ),
    warm_start( xs, array1d( -3..-2, [ 6..8, 5..7 ] ) )
  ] )
  :: seq_search( [
    warm_start_array( [                      %%% Now included in seq_search
to keep order
      warm_start( x, [<>,5,2] ),             %%% Repeated warm_starts
allowed but not specified
      warm_start( xf, array1d(-5..-3, [5.6,<>,4.7] ) ),
      warm_start( xs, array1d( -3..-2, [ 6..8, 5..7 ] ) )
    ] ),
    warm_start( [b], [true] ),
    int_search(x, first_fail, indomain_min, complete)
  ] )
  minimize x[1] + b + xf[2] + card( xs[1] intersect xs[3] );

Best regards,

On Thu, 21 Feb 2019 at 18:14, Laurent Perron <notifications@github.com>
wrote:

> I just need the spec of the flatzinc format, and some tests :-) plus some
> time.
> Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68
> 53
> 00
>
>
>
> Le jeu. 21 févr. 2019 à 04:55, Lalith Suresh <notifications@github.com> a
> écrit :
>
> > Any plans to support warm_start annotations for the CP-SAT backend? I'd
> be
> > happy to test this out if there's any early code available for this.
> >
> > —
> > You are receiving this because you were assigned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/google/or-tools/issues/539#issuecomment-465852558>,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AKj17Rs2T76JDNm4znmlLwAzvELIcbZ4ks5vPhjOgaJpZM4Q3W-k
> >
> > .
> >
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/or-tools/issues/539#issuecomment-465888146>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AQgXZn7SGrrxt7yQpG9KHNQw9-nz_mGtks5vPkdOgaJpZM4Q3W-k>
> .
>


-- 
Dr Gleb Belov       Monash University      +61 3 9903 1622

**Comment by rlloretb** [comment]
When I warm start CPLEX in Minizinc IDE, my suboptimal solution is displayed the first one. 
When I warmstart with CP-SAT, my solution is not output at all. 
On the other hand, there is no "Warning, ignored search annotation: warm_start_array(" message being output, as when I try to warmstart geocode solver, which is not compatible with warmstart in MiniZinc.
Am I missing anything?
I am using Or-tools 9.2 and MiniZinc IDE 2.6.1

**Comment by lperron** [comment]
I still do not know how to parse warm start annotations in flatzinc.
I have no minizinc example for them.
I do not know how they will be expanded to flatzinc.
This is sad, because CP-SAT supports solution hinting.

Any help is welcome.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Comment by lperron** [comment]
When I use the following code: 

```
array[1..3] of var 0..10: x;
var bool: b;
constraint b+sum(x)==1;
solve
  :: seq_search( [
    warm_start_array( [
      warm_start( x, [<>,5,2] ),
    ] ),
    warm_start( [b], [true] ),
    int_search(x, first_fail, indomain_min, complete)
  ] )
  minimize x[1] + b;
```

And I try to convert to fzn, I get:

```
/usr/local/google/home/lperron/Work/mzn/warm_start.mzn:7.7-31:
MiniZinc: type error: no function or predicate with this signature found: `warm_start(array[int] of var int,array[int] of opt int)'
Cannot use the following functions or predicates with the same identifier:
annotation warm_start(array[int] of var set of int: x,array[int] of set of int: v);
    (argument 1 expects type array[int] of var set of int, but type array[int] of var int given)
    (argument 2 expects type array[int] of set of int, but type array[int] of opt int given)
annotation warm_start(array[int] of var bool: x,array[int] of bool: v);
    (argument 1 expects type array[int] of var bool, but type array[int] of var int given)
    (argument 2 expects type array[int] of bool, but type array[int] of opt int given)
annotation warm_start(array[int] of var int: x,array[int] of int: v);
    (argument 2 expects type array[int] of int, but type array[int] of opt int given)
annotation warm_start(array[int] of var float: x,array[int] of float: v);
    (argument 2 expects type array[int] of float, but type array[int] of opt int given)
```

**Reply by rlloretb** [reply]
Does this simple example help?
If the problem is in partial initial solutions due to the <> parsing, maybe just focusing first on full initial solutions would be meaningful first step.

Minizinc:
```
int: N = 10;
int: C = 3;
array[1..N] of var 0..1: x;
constraint (sum(x) <= C);
solve :: warm_start(x, [ 1 | _ in 1..N ]) minimize sum(x);
output [show(x)];
```

In flatzinc:

```
array [1..10] of int: X_INTRODUCED_11_ = [1,1,1,1,1,1,1,1,1,1];
var 0..1: X_INTRODUCED_0_;
var 0..1: X_INTRODUCED_1_;
var 0..1: X_INTRODUCED_2_;
var 0..1: X_INTRODUCED_3_;
var 0..1: X_INTRODUCED_4_;
var 0..1: X_INTRODUCED_5_;
var 0..1: X_INTRODUCED_6_;
var 0..1: X_INTRODUCED_7_;
var 0..1: X_INTRODUCED_8_;
var 0..1: X_INTRODUCED_9_;
var 0..10: X_INTRODUCED_10_:: is_defined_var;
array [1..10] of var int: x:: output_array([1..10]) = [X_INTRODUCED_0_,X_INTRODUCED_1_,X_INTRODUCED_2_,X_INTRODUCED_3_,X_INTRODUCED_4_,X_INTRODUCED_5_,X_INTRODUCED_6_,X_INTRODUCED_7_,X_INTRODUCED_8_,X_INTRODUCED_9_];
constraint int_lin_le(X_INTRODUCED_11_,[X_INTRODUCED_0_,X_INTRODUCED_1_,X_INTRODUCED_2_,X_INTRODUCED_3_,X_INTRODUCED_4_,X_INTRODUCED_5_,X_INTRODUCED_6_,X_INTRODUCED_7_,X_INTRODUCED_8_,X_INTRODUCED_9_],3);
constraint int_lin_eq([1,1,1,1,1,1,1,1,1,1,-1],[X_INTRODUCED_0_,X_INTRODUCED_1_,X_INTRODUCED_2_,X_INTRODUCED_3_,X_INTRODUCED_4_,X_INTRODUCED_5_,X_INTRODUCED_6_,X_INTRODUCED_7_,X_INTRODUCED_8_,X_INTRODUCED_9_,X_INTRODUCED_10_],0):: ctx_neg:: defines_var(X_INTRODUCED_10_);
solve :: warm_start(x,X_INTRODUCED_11_) minimize X_INTRODUCED_10_;
```

**Comment by lperron** [comment]
Actually, I understood yesterday that I need to add some definition in
redefinition.mzn just to see the annotations in the generated fzn.


Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Reply by rlloretb** [reply]
Cool. Could it be done for version 9.3?

**Comment by lperron** [comment]
I just finished the code.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Comment by lperron** [comment, marked as answer]
And it is pushed on master.
Note that it requires minizinc 2.6.0

**Comment by lperron** [comment]
Run with -l ?

**Comment by lperron** [comment]
And the hinted solution is not feasible

**Comment by lperron** [comment]
It works for me with a valid hint:

File /usr/local/google/home/lperron/Work/fzn/warm_start.fzn parsed in 0 ms

Presolve model
  - done in 0 ms

Model warm_start
  - int_lin_eq: 1
  - int_lin_le: 1
  - Minimization problem

The number of search workers, is not specified. For better performances,
please set the number of workers to 8, 16, or more depending on the number
of cores of your computer.

Starting CP-SAT solver v9999.0.0
Parameters: log_search_progress: true search_branching: FIXED_SEARCH
num_search_workers: 1 keep_all_feasible_solutions_in_presolve: true
log_to_stdout: false max_domain_size_when_encoding_eq_neq_constraints: 32

Initial optimization model 'warm_start':
#Variables: 11 (#ints:1 in objective)
  - 10 in [0,1]
  - 1 in [0,10]
#kLinearN: 2 (#terms: 21)

Starting presolve at 0.00s
[ExtractEncodingFromLinear] #potential_supersets=0 #potential_subsets=0
#at_most_one_encodings=0 #exactly_one_encodings=0 #unique_terms=0
#multiple_terms=0 #literals=0 time=1.001e-06s
[Probing] deterministic_time: 1.8e-06 (limit: 1) wall_time: 4.496e-05
(10/10)
[DetectDuplicateConstraints] #duplicates=1 time=2.0077e-05s
[DetectDominatedLinearConstraints] #relevant_constraints=0 #work_done=0
#num_inclusions=0 #num_redundant=0 time=7.35e-07s
[DetectOverlappingColumns] #processed_columns=0 #work_done=0
#nz_reduction=0 time=1.395e-06s
[ProcessSetPPC] #relevant_constraints=0 #num_inclusions=0 work=0
time=1.022e-06s

Presolve summary:
  - 0 affine relations were detected.
  - rule 'duplicate: linear constraint parallel to objective' was applied 1
time.
  - rule 'duplicate: removed constraint' was applied 1 time.
  - rule 'duplicate: updated objective domain' was applied 1 time.
  - rule 'linear: simplified rhs' was applied 1 time.
  - rule 'linear: singleton column define objective.' was applied 1 time.
  - rule 'presolve: 0 unused variables removed.' was applied 1 time.
  - rule 'presolve: iteration' was applied 1 time.

Presolved optimization model 'warm_start':
#Variables: 10 (#bools:10 in objective)
  - 10 in [0,1]

Preloading model.
[Symmetry] Graph for symmetry has 10 nodes and 0 arcs.
[Symmetry] Symmetry computation done. time: 9.398e-06 dtime: 6e-07
#Bound   0.00s best:inf   next:[0,3]      initial_domain

Starting to load the model at 0.00s
[Probing] deterministic_time: 1.6e-06 (limit: 1) wall_time: 3.2923e-05
(10/10)
Initial num_bool: 10

Starting sequential search at 0.00s
#1       0.00s best:3     next:[0,2]       [hint] fixed_bools:0/10
#2       0.00s best:0     next:[]          fixed_bools:0/10
#Done    0.00s


CpSolverResponse summary:
status: OPTIMAL
objective: 0
best_bound: 0
booleans: 10
conflicts: 0
branches: 33
propagations: 7
integer_propagations: 55
restarts: 22
lp_iterations: 0
walltime: 0.00141427
usertime: 0.00141447
deterministic_time: 5.136e-06
gap_integral: 4.37497e-06
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Comment by lperron** [comment]
Please provide the text, I cannot even copy paste the code :-(

Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Comment by lperron** [comment]
With a correct hint:

int: N = 10;
int: C = 3;
array[1..N] of var 0..1: x;

constraint (sum(x) <= C);
solve :: warm_start(x, [1, 0, 0, 1, 0, 0, 1, 0, 0, 0]) satisfy;
output [show(x)]

outputs:

```
x = array1d(1..10, [1, 0, 0, 1, 0, 0, 1, 0, 0, 0]);
----------
==========
```
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00
